enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,191 @@
+   If no names are provided, the implicit names will be extracted from the certificates.
+1. `certPath` -  certificate full path.
+1. `keyPath` -   certificate key full path.
+
+certificate files will be read from their conifgured location by the Microshift service,
+each certification file will be validated (see validation rules).
+
+Those certificate will extend the default  API server [external](https://github.com/openshift/microshift/blob/main/pkg/controllers/kube-apiserver.go#L194) certificate configuration.
+
+
+### Workflow Description
+
+#### Default API Certs
+1. by default when there is no namedCertificates configuration the behaviour will remain the same.
+
+#### when custom namedCertificates configured
+1. Device Administrator copies the certificated to MicroShift host
+1. Device Administrator configures `namedCertificates` in the Microshift configuration yaml file (/etc/microshift/config.yaml).
+1. Device Administrator start/restarts MicroShift
+1. During startup the certifates paths will be checked and validated. in case of an error service will produce clear error log and exit.
+1. Certifates will be passed to the tls-sni-cert-key command line flag preceding all the other certificates.
+1. kube-apiserver picks up the certificates and start serving them on the configured SNI.
+
+### Implementation Details
+
+The certs will be prepended to the []configv1.NamedCertificate list before the api server is started. (it will be added to `-tls-sni-cert-key` flag)
+
+this certification paths configuration and names will be prepended into the kube-apiserver `tls-sni-cert-key` command line flag.
+when same SNI appear in the CN part of the provided certs,this certificate will take precedence over the `default` external-signer. [ref](https://github.com/kubernetes/kubernetes/blob/98358b8ce11b0c1878ae7aa1482668cb7a0b0e23/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go#L38)
+
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+* User provide incorrect/expired certificate.
+  kube-api server will continue serving with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+* User provided expired Certificate and Microshift service started/restarted 
+  Microshift will start with a warning in the logs,kube-api will continue serve with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+### Drawbacks
+* External certificate wont be automaticly renewed, so it requires manual Certificate rotation. 
+  similiar approach is taken by OpenShift.
+
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+N/A
+
+### Test Plan
+add e2e test that will:
+- generate and sign certificates with custom ca.
+- change the default Microshift configuration to use the newly generated certs.
+- make sure system is functional using  generated external kubeconfig.
```

Comment:
its infeasible to generate kubeconfig for every provided option in NameCert (we also support wildcards ie)
IMO  Microshift should generate one kubeconfig for all the ,without `certificate-authority-data`,
it might be in-accurate kubeconfig (since its hard to guess the SNI..) or just document how to change the kubeconfig manually .

Reply:
Not verifying the validity of the server cert (omitting `certificate-authority-data`) is almost as bad as connecting with `--insecure-skip-tls-verify`. It exposes the connection to MitM attacks. Additionally, I don't think that someone who is taking the trouble to specify their own cert for securing the apiserver will be interested in permanently ignoring cert validation.

Reply:
from my experiments omitting certificate-authority-data from the kubeconfig will  cause the `oc` client to  use  the  operating system  trusted CAs ( from RHEL trustore)  (instead specifying the exact CA).  so it disables nothing.

Reply:
Ii've  performed similar experiment with regular OCP cluster, documented it [here.](https://gist.github.com/eslutsky/898a034e3da043ad013bcfbcdb911705)